src/map/elements/crab.rs

```diff
@@ -0,0 +1,204 @@
+    >,
+    element_assets: Res<Assets<MapElementMeta>>,
+) {
+    // Hydrate the newly spawned crabs
+    for (entity, map_element_meta, initial_pos) in &non_hydrated_map_elements {
+        let map_element = element_assets.get(map_element_meta).unwrap();
+        if let BuiltinElementKind::Crab {
+            start_frame,
+            end_frame,
+            fps,
+            atlas_handle,
+            comfortable_spawn_distance,
+            comfortable_scared_distance,
+            same_level_threshold,
+            walk_speed,
+            run_speed,
+            ..
+        } = &map_element.builtin
+        {
+            let config = CrabConfig {
+                comfortable_spawn_distance: *comfortable_spawn_distance,
+                comfortable_scared_distance: *comfortable_scared_distance,
+                same_level_threshold: *same_level_threshold,
+                walk_speed: *walk_speed,
+                run_speed: *run_speed,
+            };
+
+            commands
+                .entity(entity)
+                .insert(MapElementHydrated)
+                .insert(Name::new("Environment: Crab"))
+                .insert(AnimatedSprite {
+                    start: *start_frame,
+                    end: *end_frame,
+                    atlas: atlas_handle.inner.clone(),
+                    repeat: true,
+                    fps: *fps,
+                    ..default()
+                })
+                .insert(CrabCritter {
+                    state: CrabState::Walking { left: false },
+                    state_timer: Timer::default(),
+                    config,
+                    start_pos: Vec2::new(initial_pos.translation.x, initial_pos.translation.y),
+                })
+                .insert(KinematicBody {
+                    size: Vec2::new(17.0, 12.0),
+                    offset: Vec2::new(0.0, 0.0),
+                    gravity: 1.0,
+                    has_mass: true,
+                    has_friction: true,
+                    ..default()
+                });
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum CrabState {
+    Paused,
+    Walking { left: bool },
+    Fleeing { scared_of: Vec2 },
+}
+
+impl CrabState {
+    fn is_moving(&self) -> bool {
+        matches!(self, CrabState::Walking { .. } | CrabState::Fleeing { .. })
+    }
+}
+
+pub struct CrabConfig {
+    comfortable_spawn_distance: f32,
+    comfortable_scared_distance: f32,
+    same_level_threshold: f32,
+    walk_speed: f32,
+    run_speed: f32,
+}
+
+#[derive(Component)]
+pub struct CrabCritter {
+    state: CrabState,
+    start_pos: Vec2,
+    config: CrabConfig,
+    state_timer: Timer,
+}
+
+fn update_crabs(
+    mut crab_query: Query<(
+        &mut CrabCritter,
+        &mut AnimatedSprite,
+        &Transform,
+        &mut KinematicBody,
+    )>,
+    scary_things_query: Query<&Transform, (With<PlayerIdx>, Without<CrabCritter>)>,
+    time: Res<Time>,
+) {
+    let mut rng_bool = rand::thread_rng();
```

Comment:
For the random generation, try using the `GlobalRng` resource from the `crate::random` module.

That should be faster, and make sure that different computers in a network game generate the same random numbers.